dataset/test/assign_test.cpp

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+#include <gtest/gtest.h>
+
+#include "scipp/dataset/dataset.h"
+#include "scipp/dataset/except.h"
+
+using namespace scipp;
+using namespace scipp::dataset;
+
+struct AssignTest : public ::testing::Test {
+  AssignTest() {}
+
+protected:
+  Dimensions dims{Dim::X, 3};
+  Variable data = makeVariable<double>(dims, Values{1, 2, 3});
+  Variable x = makeVariable<double>(dims, Values{1, 1, 3});
+  Variable mask = makeVariable<bool>(dims, Values{true, false, true});
+  DataArray array{data, {{Dim::X, copy(x)}}, {{"mask", mask}}};
+};
+
+TEST_F(AssignTest, self) {
+  const auto original = copy(array);
+  EXPECT_EQ(array.assign(array), original);
+}
+
+TEST_F(AssignTest, coord_fail) {
+  const auto original = copy(array);
+  EXPECT_THROW(array.assign(array.slice({Dim::X, 0, 1})),
+               except::CoordMismatchError);
+  EXPECT_EQ(array, original);
+  EXPECT_THROW(array.slice({Dim::X, 0, 1}).assign(array.slice({Dim::X, 2, 3})),
+               except::CoordMismatchError);
+  EXPECT_EQ(array, original);
+}
+
+TEST_F(AssignTest, mask_propagation) {
+  const auto original = copy(array);
+  // Mask values get copied
+  array.slice({Dim::X, 0}).assign(original.slice({Dim::X, 1}));
+  EXPECT_EQ(array.masks()["mask"],
+            makeVariable<bool>(dims, Values{false, false, true}));
+  array.slice({Dim::X, 0}).assign(original.slice({Dim::X, 2}));
+  EXPECT_EQ(array.masks()["mask"],
+            makeVariable<bool>(dims, Values{true, false, true}));
+  // Mask not in source is preserved unchanged
+  array.masks().set("other", mask);
+  array.slice({Dim::X, 0}).assign(original.slice({Dim::X, 1}));
+  EXPECT_EQ(array.masks()["mask"], mask);
```

Comment:
Is this correct? Shouldn't this be `EXPECT_EQ(array.masks()["other"], mask);`?

Reply:
Actually, why is `"mask"` preserved?

Reply:
Hmm, good question:

- If `self` is a slice, we cannot remove the mask.
- Consider:
  ```python
  data['spectrum' , 17] = spectrum
  ```
  If `data` has a `tof` mask but `spectrum` does not, we should not remove the mask, since it as a mask for *all* spectra?

I feel these arguments favor keeping the mask, but I agree that it is not 100% clear. Are there different relevant examples I did not list?

Reply:
I think I answered the wrong question, again, what I wrote is why `other` is preserved...

Reply:
My second point was that this looks like a bug. `"mask"` should have been changed by the assignment regardless of any other masks.